enhancements/two-nodes-openshift/2no_template.md

```diff
@@ -0,0 +1,589 @@
+   4. Pacemaker on Node1 forces E-quorum (etcd promotion event)
+   5. Cluster continues with no redundancy
+   6. … time passes …
+   7. Node2 has a persistent failure that prevents communication with Node1
+      * Node2 does not have C-quorum (requires forming a membership with it’s peer)
+      * Node2 does not start etcd or kubelet, remains inert waiting for Node1
+   8. Persistent failure on Node2 is repaired
+   9. Corosync membership containing both nodes forms
+   10. Pacemaker “starts” etcd on Node2 as a follower of Node1
+   11. Pacemaker “promotes” etcd on Node2 as full replica of Node1
+   12. Pacemaker starts kubelet
+   13. Cluster continues with 1+1 redundancy
+4. Two Failures
+   1. Node2 failure (1st failure)
+   2. Corosync on the survivor (Node1)
+   3. Etcd loses internal quorum (E-quorum) and goes read-only
+   4. Node1 retains “corosync quorum” (C-quorum) and initiates fencing of Node2
+   5. Pacemaker on Node1 forces E-quorum (etcd promotion event)
+   6. Cluster continues with no redundancy
+   7. Node1 experience a power failure (2nd Failure)
+   8. … time passes …
+   9. Node1 Power restored
+   10. Node1 boots but can not gain quorum before Node2 joins the cluster due to a risk of fencing loop
+       * Mitigation (Phase 1): manual intervention (possibly a script)  in case admin can guarantee Node2 is down, which will grant Node1 quorum and restore cluster limited (none HA) functionality.
+       * Mitigation (Phase 2): limited automatic intervention for some use cases: for example Node1 will gain quorum only if Node2 can be verified to be down by successfully querying its BMC status.
+5. Kubelet Failure
+   1. Pacemaker’s monitoring detects the failure
+   2. Pacemaker restarts kubelet
+   3. Stop failure is optionally escalated to a node failure (fencing)
+   4. Start failure defaults to leaving the service offline
+6. Etcd Failure
+   1. Pacemaker’s monitoring detects the failure
+   2. Pacemaker either demotes etcd so it can resync, or restarts and promotes etcd
+   3. Stop failure is optionally escalated to a node failure (fencing)
+   4. Start failure defaults to leaving the service offline
+
+
+### API Extensions
+
+API Extensions are CRDs, admission and conversion webhooks, aggregated API servers,
+and finalizers, i.e. those mechanisms that change the OCP API surface and behaviour.
+
+- Name the API extensions this enhancement adds or modifies.
+- Does this enhancement modify the behaviour of existing resources, especially those owned
+  by other parties than the authoring team (including upstream resources), and, if yes, how?
+  Please add those other parties as reviewers to the enhancement.
+
+  Examples:
+  - Adds a finalizer to namespaces. Namespace cannot be deleted without our controller running.
+  - Restricts the label format for objects to X.
+  - Defaults field Y on object kind Z.
+
+Fill in the operational impact of these API Extensions in the "Operational Aspects
+of API Extensions" section.
```

Comment:
>something that can signal CEO 

I think that we prefer something "general" which can be used by other stakeholder process (instead of creating multiple localized flags).

Reply:
What about introducing a feature flag for this? Can be read by all components which need to deal with 2NO, and can be set in install config AFAIK.

Reply:
we need to signal _after_ the installation is completed

Reply:
> we need to signal after the installation is completed

Hmm I'm not sure that's an issue, can't we turn on the flag as part of the installation process ? (i.e after the originally bootstrapped node has joined the cluster as a regular node)

Reply:
You guys need to signal after all your stuff is ready to run etcd, is there a guarantee that this is true after the bootstrap node has joined the cluster?

I was under the impression there is a fair bit of installation involved.

Reply:
>is there a guarantee that this is true after the bootstrap node has joined the cluster as a real control plane node?

The way I understand it: past that point we can set up pacemaker/corosync (so it would be after the node joining the cluster).
IIUC in this particular context it's not very important where (timing wise) is the point in which we can signal CEO to relinquish control of etcd, as to the fact that there is such a point in time (in which we'll turn the flag).

Reply:
That's something you should write down in this enhancement, in which case you're accepting some amount of downtime of the API during the installation. Not sure if there's also race conditions involved where podman is coming up faster than the static etcd pod and we're in an eternal static pod revision rollout crash loop.

Note that a bunch of CI jobs will not going to like this kind of disruption.

Reply:
Doing a short recap following our last sync:
- Since we want CEO to relinquish etcd control to pacemaker after bootstrapping is complete we can't relay on the installer to do that.
- One potential option is using LCO (Local Cluster Operator) to somehow flag CEO to relinquish control of etcd.
- What kind of flag we will use to trigger MCO is still TBD

Let me know if that makes sense.

Reply:
Resolved